src/controller/python/chip/bdx/bdx-transfer.cpp

```diff
@@ -0,0 +1,220 @@
+    }
+}
+
+CHIP_ERROR BdxTransfer::AcceptSend()
+{
+    VerifyOrReturnError(mAwaitingAccept, CHIP_ERROR_INCORRECT_STATE);
+    mAwaitingAccept = false;
+
+    TransferSession::TransferAcceptData acceptData;
+    acceptData.ControlMode  = TransferControlFlags::kSenderDrive;
+    acceptData.MaxBlockSize = mTransfer.GetTransferBlockSize();
+    acceptData.StartOffset  = mTransfer.GetStartOffset();
+    acceptData.Length       = mTransfer.GetTransferLength();
+    return mTransfer.AcceptTransfer(acceptData);
+}
+
+CHIP_ERROR BdxTransfer::AcceptReceive(const ByteSpan & data_to_send)
+{
+    VerifyOrReturnError(mAwaitingAccept, CHIP_ERROR_INCORRECT_STATE);
+    mAwaitingAccept = false;
+    mData           = new uint8_t[data_to_send.size()];
+    if (mData == nullptr)
+    {
+        return CHIP_ERROR_NO_MEMORY;
+    }
+    memcpy(mData, data_to_send.data(), data_to_send.size());
+    mDataCount = data_to_send.size();
+
+    TransferSession::TransferAcceptData acceptData;
+    acceptData.ControlMode  = TransferControlFlags::kReceiverDrive;
+    acceptData.MaxBlockSize = mTransfer.GetTransferBlockSize();
+    acceptData.StartOffset  = mTransfer.GetStartOffset();
+    acceptData.Length       = mTransfer.GetTransferLength();
+    return mTransfer.AcceptTransfer(acceptData);
+}
+
+CHIP_ERROR BdxTransfer::Reject()
+{
+    VerifyOrReturnError(mAwaitingAccept, CHIP_ERROR_INCORRECT_STATE);
+    mAwaitingAccept = false;
+    return mTransfer.RejectTransfer(StatusCode::kTransferFailedUnknownError);
+}
+
+void BdxTransfer::HandleTransferSessionOutput(TransferSession::OutputEvent & event)
+{
+    assertChipStackLockedByCurrentThread();
+
+    ChipLogDetail(BDX, "Received event %s", event.ToString(event.EventType));
+
+    switch (event.EventType)
+    {
+    case TransferSession::OutputEventType::kInitReceived:
+        mAwaitingAccept = true;
+        mDelegate->InitMessageReceived(this, event.transferInitData);
+        break;
+    case TransferSession::OutputEventType::kStatusReceived:
+        ChipLogError(BDX, "Received StatusReport %x", static_cast<uint16_t>(event.statusData.statusCode));
+        EndSession(CHIP_ERROR_INTERNAL);
+        break;
+    case TransferSession::OutputEventType::kInternalError:
+        EndSession(CHIP_ERROR_INTERNAL);
+        break;
+    case TransferSession::OutputEventType::kTransferTimeout:
+        EndSession(CHIP_ERROR_TIMEOUT);
+        break;
+    case TransferSession::OutputEventType::kBlockReceived:
+        if (mDelegate)
+        {
+            ByteSpan data(event.blockdata.Data, event.blockdata.Length);
+            mDelegate->DataReceived(this, data);
+            mTransfer.PrepareBlockAck();
+        }
+        break;
+    case TransferSession::OutputEventType::kMsgToSend:
+        SendMessage(event);
+        if (event.msgTypeData.HasMessageType(MessageType::BlockAckEOF))
+        {
+            // TODO: Ending the session here means the StandaloneAck for the BlockAckEOF message hasn't been received.
```

Comment:
Is this going to cause similar flow control problems to the apple yaml tests where the ack is assumed to be for something else? Or does the exchange get killed?

Reply:
This can cause flakiness, but in a different way than the flakiness seen in the yaml tests: the test runner can continue and not let the test device end its BDX session. The diagnostic logs cluster will reject log requests until the session has ended. I saw this when I was investigating the flaky tests, which is why I've added a TODO here. I'm not sure if there's a way to know when the ack has been received.

Reply:
ok, let's leave this for now then. It sounds like the diagnostic logs cluster properly handles the dangling sessions in some way. There's probably SOMETHING that can be used to block, but if it's not going to cause flakes I think leaving as a todo is fine